wordpress-mobile/WordPress-FluxC-Android
# Add Observable totalDuration to PagedListWrapper

Part of https://github.com/woocommerce/woocommerce-android/issues/3568. 

## What

I added a new `listChanged` observable in `PagedListWrapper` that emits `totalDuration` values. This will be tracked in WC to help gather more information for the slow orders reports (ref p91TBi-4p5-p2). This will also help confirm if whatever we're going to _optimize_ in the future worked. 

https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/4c8c08780a413f21d529f6a6c9cc175229e9af7d/fluxc/src/main/java/org/wordpress/android/fluxc/model/list/PagedListWrapper.kt#L46-L54

https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/4c8c08780a413f21d529f6a6c9cc175229e9af7d/fluxc/src/main/java/org/wordpress/android/fluxc/model/list/PagedListWrapper.kt#L73-L74

### Order IDs Fetching Only

This only tracks the duration of loading the order list summaries via [`FETCH_ORDER_LIST`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/plugins/woocommerce/src/main/java/org/wordpress/android/fluxc/action/WCOrderAction.java#L41-L42) / [`fetchOrderListSummaries`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/network/rest/wpcom/wc/order/OrderRestClient.kt#L123-L136). I thought about including the full `Order` page loads as a whole ([`FETCH_ORDER_BY_IDS`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/plugins/woocommerce/src/main/java/org/wordpress/android/fluxc/action/WCOrderAction.java#L44-L45)), but it looks like that will result in **incomparable total durations** because we only fetch orders that have not been previously loaded. Example: 

https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/example/src/main/java/org/wordpress/android/fluxc/example/WCOrderListActivity.kt#L233-L239

So, I think it's better to track consistent fetches like the `fetchOrderListSummaries`. We will only be using the Tracks event to have an idea of how we're progressing as we make changes to the code base and/or API to improve the queries. So I think this will be enough. 

### Total Duration Calculation


![image](https://user-images.githubusercontent.com/198826/109230347-f338bb00-7781-11eb-9580-135c2d1b818f.png)


The total duration calculation is achieved by storing the `requestStartTime` in [`FetchOrderListPayload`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt#L50-L54) and passing it through [`FetchOrderListResponsePayload`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt#L71-L85). `FetchOrderListResponsePayload` eventually leads to dispatching `FETCHED_LIST_ITEMS` with a [`FetchedListItemsPayload`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/plugins/woocommerce/src/main/kotlin/org/wordpress/android/fluxc/store/WCOrderStore.kt#L573-L583). `ListStore` will [receive this event](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/fluxc/src/main/java/org/wordpress/android/fluxc/store/ListStore.kt#L261-L265), **calculate** the total duration, and emit an `OnListChanged`. This is received by `PagedListWrapper` which [emits a value through the `listChanged` `LiveData`](https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/fluxc/src/main/java/org/wordpress/android/fluxc/model/list/PagedListWrapper.kt#L144) that we can observe and add the tracking in: 

 https://github.com/wordpress-mobile/WordPress-FluxC-Android/blob/06b9aed61a63105839b2a4656d242e3f19d8a9f5/example/src/main/java/org/wordpress/android/fluxc/example/WCOrderListActivity.kt#L146-L151


## Testing

1. Please use the branch in https://github.com/woocommerce/woocommerce-android/pull/3635
2. Log in and navigate to the Orders list → Processing. Pull to refresh.
3. Confirm that there is a log like this:
    ```
    🔵 Tracked: orders_list_loaded, Properties: {"total_duration":1.148,"status":"processing","blog_id":171751943,"is_wpcom_store":false}
    ```
4. Navigate to All Orders. Pull to refresh.
5. Confirm that there is a log like this:
    ```
    🔵 Tracked: orders_list_loaded, Properties: {"total_duration":1.148,"status":null,"blog_id":171751943,"is_wpcom_store":false}
    ```
5. Scroll down so the next set of Orders is loaded. Make sure you have lots of orders though. 😅 
5. Confirm that there is a log like this:
    ```
    🔵 Tracked: orders_list_loaded, Properties: {"total_duration":1.148,"status":null,"blog_id":171751943,"is_wpcom_store":false}
    ```
6. Tap on Search and pick the Completed status. 
5. Confirm that there is a log like this:
    ```
    🔵 Tracked: orders_list_loaded, Properties: {"total_duration":1.148,"status":"completed","blog_id":171751943,"is_wpcom_store":false}
    ```

## Author Checklist
